### ISSUE TITLE
SQL-Group segmentation fault

### ISSUE THREAD

**Issue by pyrrhudite** [FreeRADIUS/freeradius-server#1657]
# Issue type
- [x] Defect - Crash or memory corruption.

# Defect/Feature description
## How to reproduce issue

       update request {
                       Huntgroup-Name := "%{sql:SELECT groupname FROM radhuntgroup WHERE nasipaddress='%{NAS-IP-Address}'}"
        }

        if (SQL-Group == "Profile_Beta_User") {
               if (Huntgroup-Name == "tier3") {
                       reject
               }
        }

If the client is not a member of "Profile_Beta_User", radiusd will crash. Otherwise everything is cool.
 
## Output of ``[radiusd|freeradius] -X`` showing issue ocurring
```
(1) sql: EXPAND %{User-Name}
(1) sql:    --> test.radius@gmail.com
(1) sql: SQL-User-Name set to 'test.radius@gmail.com'
rlm_sql (sql): Reserved connection (1)
(1) sql: EXPAND SELECT id, username, attribute, value, op FROM radcheck WHERE username = '%{SQL-User-Name}' ORDER BY id
(1) sql:    --> SELECT id, username, attribute, value, op FROM radcheck WHERE username = 'test.radius@gmail.com' ORDER BY id
(1) sql: Executing select query: SELECT id, username, attribute, value, op FROM radcheck WHERE username = 'test.radius@gmail.com' ORDER BY id
(1) sql: User found in radcheck table
(1) sql: Conditional check items matched, merging assignment check items
(1) sql:   Cleartext-Password := "**.**"
(1) sql:   Simultaneous-Use := 2
(1) sql:   Expiration := "Aug 10 2016 00:00:00 CST"
(1) sql: EXPAND SELECT id, username, attribute, value, op FROM radreply WHERE username = '%{SQL-User-Name}' ORDER BY id
(1) sql:    --> SELECT id, username, attribute, value, op FROM radreply WHERE username = 'test.radius@gmail.com' ORDER BY id
(1) sql: Executing select query: SELECT id, username, attribute, value, op FROM radreply WHERE username = 'test.radius@gmail.com' ORDER BY id
(1) sql: EXPAND SELECT groupname FROM radusergroup WHERE username = '%{SQL-User-Name}' ORDER BY priority
(1) sql:    --> SELECT groupname FROM radusergroup WHERE username = 'test.radius@gmail.com' ORDER BY priority
(1) sql: Executing select query: SELECT groupname FROM radusergroup WHERE username = 'test.radius@gmail.com' ORDER BY priority
(1) sql: User found in the group table
(1) sql: EXPAND SELECT id, groupname, attribute, Value, op FROM radgroupcheck WHERE groupname = '%{SQL-Group}' ORDER BY id
(1) sql:    --> SELECT id, groupname, attribute, Value, op FROM radgroupcheck WHERE groupname = 'Profile_Monthly_Traffic_A' ORDER BY id
(1) sql: Executing select query: SELECT id, groupname, attribute, Value, op FROM radgroupcheck WHERE groupname = 'Profile_Monthly_Traffic_A' ORDER BY id
(1) sql: Group "Profile_Monthly_Traffic_A": Conditional check items matched
(1) sql: Group "Profile_Monthly_Traffic_A": Merging assignment check items
(1) sql:   Max-Monthly-Traffic := 26843545600
(1) sql: EXPAND SELECT id, groupname, attribute, value, op FROM radgroupreply WHERE groupname = '%{SQL-Group}' ORDER BY id
(1) sql:    --> SELECT id, groupname, attribute, value, op FROM radgroupreply WHERE groupname = 'Profile_Monthly_Traffic_A' ORDER BY id
(1) sql: Executing select query: SELECT id, groupname, attribute, value, op FROM radgroupreply WHERE groupname = 'Profile_Monthly_Traffic_A' ORDER BY id
(1) sql: Group "Profile_Monthly_Traffic_A": Merging reply items
rlm_sql (sql): Released connection (1)
(1)     [sql] = ok
(1)     update request {
(1)       EXPAND %{User-Name}
(1)          --> test.radius@gmail.com
(1)       SQL-User-Name set to 'test.radius@gmail.com'
rlm_sql (sql): Reserved connection (2)
(1)       Executing select query: SELECT groupname FROM radhuntgroup WHERE nasip address='xx.xx.xx.xxx'
rlm_sql (sql): Released connection (2)
(1)       EXPAND %{sql:SELECT groupname FROM radhuntgroup WHERE nasipaddress='%{NAS-IP-Address}'}
(1)          --> tier3
(1)       Huntgroup-Name := tier3
(1)     } # update request = noop
(1)     if (SQL-Group == "Profile_Beta_User") {
(1)     sql_groupcmp
(1)     EXPAND %{User-Name}
(1)        --> test.radius@gmail.com
(1)     SQL-User-Name set to 'test.radius@gmail.com'
rlm_sql (sql): Reserved connection (0)
(1)     EXPAND SELECT groupname FROM radusergroup WHERE username = '%{SQL-User-Name}' ORDER BY priority
(1)        --> SELECT groupname FROM radusergroup WHERE username = 'test.radius@gmail.com' ORDER BY priority
(1)     Executing select query: SELECT groupname FROM radusergroup WHERE username = 'test.radius@gmail.com' ORDER BY priority
rlm_sql (sql): Released connection (0)
(1)     sql_groupcmp finished: User is NOT a member of group Profile_Beta_User
Segmentation fault (core dumped
```
```text
Ubuntu 14.04 freeradius 3.0.11 ./configure --sysconfdir=/etc/ --with-udpfromto --enable-developer --with-openssl=no
```
## Full backtrace from LLDB or GDB
```
Starting program: /usr/local/sbin/radiusd -f
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff1c37700 (LWP 12636)]
[Thread 0x7ffff1c37700 (LWP 12636) exited]
[New Thread 0x7ffff1c37700 (LWP 12637)]
[New Thread 0x7ffff0b0b700 (LWP 12638)]
[New Thread 0x7fffebfff700 (LWP 12639)]





Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff0b0b700 (LWP 12638)]
paircompare (request=request@entry=0x9484e0, req_list=0x949e60,
    check=check@entry=0x7fffe4001390, rep_list=rep_list@entry=0x0)
    at src/main/pair.c:537
537                                     if ((auth_item->da == from) || (!from))                                                                                                              {
(gdb) info threads
  Id   Target Id         Frame
  5    Thread 0x7fffebfff700 (LWP 12639) "radiusd" sem_wait ()
    at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
* 4    Thread 0x7ffff0b0b700 (LWP 12638) "radiusd" paircompare (
    request=request@entry=0x9484e0, req_list=0x949e60,
    check=check@entry=0x7fffe4001390, rep_list=rep_list@entry=0x0)
    at src/main/pair.c:537
  3    Thread 0x7ffff1c37700 (LWP 12637) "radiusd" sem_wait ()
    at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
  1    Thread 0x7ffff7fec780 (LWP 12631) "radiusd" 0x00007ffff6e2dc53 in select
    () at ../sysdeps/unix/syscall-template.S:81
(gdb)  thread apply all bt full
Thread 5 (Thread 0x7fffebfff700 (LWP 12639)):
#0  sem_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
No locals.
#1  0x00000000004290af in request_handler_thread (arg=0x933da0)
    at src/main/threads.c:621
        self = 0x933da0
#2  0x00007ffff7342184 in start_thread (arg=0x7fffebfff700)
    at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7fffebfff700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737152808704,
                7779334135500148084, 1, 0, 140737152809408, 140737152808704,
                -7779290154466898572, -7779316046496494220},
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0},
            data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#3  0x00007ffff6e3637d in clone ()
---Type <return> to continue, or q <return> to quit---
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 4 (Thread 0x7ffff0b0b700 (LWP 12638)):
#0  paircompare (request=request@entry=0x9484e0, req_list=0x949e60,
    check=check@entry=0x7fffe4001390, rep_list=rep_list@entry=0x0)
    at src/main/pair.c:537
        cursor = {first = 0x7ffff0b09e38, found = 0x0, last = 0x0,
          current = 0x7fffe4001390, next = 0x0}
        check_item = 0x7fffe4001390
        auth_item = 0x656d616e70756f72
        from = 0x65d940
        result = 0
        compare = <optimized out>
#1  0x00007ffff7bb7561 in cond_cmp_values (request=0x9484e0,
    lhs_type=<optimized out>, lhs=<optimized out>, lhs_len=<optimized out>,
    rhs_type=PW_TYPE_STRING, rhs=0x7ffff0b09ff0, rhs_len=25,
    c=<optimized out>, c=<optimized out>) at src/main/evaluate.c:315
        vp = 0x7fffe4001390
        map = <optimized out>
        rcode = <optimized out>
#2  0x00007ffff7bb79d1 in cond_normalise_and_cmp (request=0x9484e0,
    c=c@entry=0x894880, lhs_type=lhs_type@entry=PW_TYPE_INVALID,
---Type <return> to continue, or q <return> to quit---
    lhs_enumv=lhs_enumv@entry=0x0, lhs=lhs@entry=0x0, lhs_len=lhs_len@entry=0)
    at src/main/evaluate.c:585
        ret = <optimized out>
        data = {strvalue = 0x894cb0 "Profile_Monthly_Traffic_A",
          octets = 0x894cb0 "Profile_Monthly_Traffic_A", integer = 8998064,
          ipaddr = {s_addr = 8998064}, date = 8998064, filter = {8998064,
            9739264, 140737347430922, 0}, ifid = "\260L\211\000\000\000\000",
          ipv6addr = {__in6_u = {
              __u6_addr8 = "\260L\211\000\000\000\000\000\000\234\224\000\000\000\000", __u6_addr16 = {19632, 137, 0, 0, 39936, 148, 0, 0}, __u6_addr32 = {
                8998064, 0, 9739264, 0}}},
          ipv6prefix = "\260L\211\000\000\000\000\000\000\234\224\000\000\000\000\000\n\252", byte = 176 '\260', ushort = 19632, ether = "\260L\211\000\000",
          sinteger = 8998064, integer64 = 8998064,
          ipv4prefix = "\260L\211\000\000", ptr = 0x894cb0}
        map = 0x8946d0
        cast = <optimized out>
        cast_type = <optimized out>
        rcode = 0
        rhs_type = PW_TYPE_STRING
        rhs_enumv = 0x0
        rhs = <optimized out>
        rhs_len = 25
---Type <return> to continue, or q <return> to quit---
        lhs_cast = {strvalue = 0x7ffff0b0a030 "\320F\211",
          octets = 0x7ffff0b0a030 "\320F\211", integer = 4038107184, ipaddr = {
            s_addr = 4038107184}, date = 4038107184, filter = {
            140737231495216, 2, 9219328, 0},
          ifid = "0\240\260\360\377\177\000", ipv6addr = {__in6_u = {
              __u6_addr8 = "0\240\260\360\377\177\000\000\002\000\000\000\000\000\000", __u6_addr16 = {41008, 61616, 32767, 0, 2, 0, 0, 0}, __u6_addr32 = {
                4038107184, 32767, 2, 0}}},
          ipv6prefix = "0\240\260\360\377\177\000\000\002\000\000\000\000\000\000\000\000\255", byte = 48 '0', ushort = 41008,
          ether = "0\240\260\360\377\177", sinteger = -256860112,
          integer64 = 140737231495216, ipv4prefix = "0\240\260\360\377\177",
          ptr = 0x7ffff0b0a030}
        rhs_cast = {strvalue = 0x7ffff0b0a0ac "", octets = 0x7ffff0b0a0ac "",
          integer = 4038107308, ipaddr = {s_addr = 4038107308},
          date = 4038107308, filter = {140737231495340, 140737294229918,
            140737231495344, 9739136}, ifid = "\254\240\260\360\377\177\000",
          ipv6addr = {__in6_u = {
              __u6_addr8 = "\254\240\260\360\377\177\000\000\236\341m\364\377\177\000", __u6_addr16 = {41132, 61616, 32767, 0, 57758, 62573, 32767, 0},
              __u6_addr32 = {4038107308, 32767, 4100841886, 32767}}},
          ipv6prefix = "\254\240\260\360\377\177\000\000\236\341m\364\377\177\000\000\260\240", byte = 172 '\254', ushort = 41132,
---Type <return> to continue, or q <return> to quit---
          ether = "\254\240\260\360\377\177", sinteger = -256859988,
          integer64 = 140737231495340,
          ipv4prefix = "\254\240\260\360\377\177", ptr = 0x7ffff0b0a0ac}
        lhs_cast_buff = <optimized out>
        rhs_cast_buff = 0x0
        escape = <optimized out>
#3  0x00007ffff7bb831c in radius_evaluate_map (request=request@entry=0x9484e0,
    modreturn=modreturn@entry=2, depth=depth@entry=0, c=c@entry=0x894880)
    at src/main/evaluate.c:654
        vp = <optimized out>
        cursor = {first = 0x9484e0, found = 0x2,
          last = 0x7ffff7dd9b28 <rad_debug_lvl>, current = 0x1,
          next = 0x8cb630}
        rcode = <optimized out>
        map = 0x8946d0
#4  0x00007ffff7bb8420 in radius_evaluate_cond (
    request=request@entry=0x9484e0, modreturn=2, depth=depth@entry=0,
    c=0x894880) at src/main/evaluate.c:750
        rcode = -1
#5  0x00000000004215e7 in modcall_recurse (request=0x9484e0,
    component=MOD_AUTHORIZE, depth=1, entry=entry@entry=0x7ffff0b0a598,
    do_next_sibling=true) at src/main/modcall.c:479
        condition = <optimized out>
---Type <return> to continue, or q <return> to quit---
        if_taken = false
        was_if = <optimized out>
        c = 0x92a140
        priority = -1
        result = RLM_MODULE_OK
#6  0x0000000000421329 in modcall_child (request=<optimized out>,
    component=<optimized out>, depth=<optimized out>, entry=0x7ffff0b0a580,
    c=<optimized out>, result=0x7ffff0b0a3e4, do_next_sibling=true)
    at src/main/modcall.c:408
        next = 0x7ffff0b0a598
#7  0x000000000042150e in modcall_recurse (request=request@entry=0x9484e0,
    component=component@entry=MOD_AUTHORIZE, depth=depth@entry=0,
    entry=entry@entry=0x7ffff0b0a580,
    do_next_sibling=do_next_sibling@entry=true) at src/main/modcall.c:785
        if_taken = false
        was_if = false
        c = 0x926aa0
        priority = -1
        result = RLM_MODULE_UNKNOWN
#8  0x0000000000422743 in modcall (component=component@entry=MOD_AUTHORIZE,
    c=c@entry=0x926aa0, request=request@entry=0x9484e0)
    at src/main/modcall.c:1130
        stack = {{result = RLM_MODULE_NOTFOUND, priority = 0, unwind = 0,
---Type <return> to continue, or q <return> to quit---
            c = 0x926aa0}, {result = RLM_MODULE_UPDATED, priority = 4,
            unwind = 0, c = 0x92a140}, {result = RLM_MODULE_NOTFOUND,
            priority = 0, unwind = 0, c = 0x0}, {result = RLM_MODULE_REJECT,
            priority = 0, unwind = 0, c = 0x0} <repeats 19 times>, {
            result = RLM_MODULE_REJECT, priority = 0, unwind = -141073162,
            c = 0x7ffff0b0a7e0}, {result = 24, priority = 0, unwind = 0,
            c = 0xffffff80}, {result = RLM_MODULE_OK, priority = 0,
            unwind = -141023732, c = 0x0}, {result = 9734048, priority = 0,
            unwind = -256858152, c = 0x949150}, {result = RLM_MODULE_REJECT,
            priority = 0, unwind = 9736528, c = 0x949270}, {result = 8972944,
            priority = 0, unwind = 6649920, c = 0x3d44941f}, {
            result = 9647024, priority = 0, unwind = -140979943,
            c = 0x9333b0}, {result = 4340636, priority = 0,
            unwind = -918514401, c = 0x42e75e9711d078d5}, {
            result = RLM_MODULE_REJECT, priority = 0, unwind = 9734048,
            c = 0x7ffff0b0a858}, {result = 4307184, priority = 0, unwind = 0,
            c = 0x9484e0}}
#9  0x000000000041d55d in indexed_modcall (comp=comp@entry=MOD_AUTHORIZE,
    idx=idx@entry=0, request=request@entry=0x9484e0) at src/main/modules.c:1028
        rcode = <optimized out>
        list = 0x926aa0
        server = <optimized out>
#10 0x000000000041e50f in process_authorize (autz_type=autz_type@entry=0,
---Type <return> to continue, or q <return> to quit---
    request=request@entry=0x9484e0) at src/main/modules.c:2123
No locals.
#11 0x000000000040cce6 in rad_authenticate (request=0x9484e0)
    at src/main/auth.c:489
        check_item = <optimized out>
        module_msg = <optimized out>
        tmp = <optimized out>
        result = <optimized out>
        autz_retry = 0 '\000'
        autz_type = 0
#12 0x00000000004301b2 in request_running (request=0x9484e0,
    action=<optimized out>) at src/main/process.c:1527
        __FUNCTION__ = "request_running"
#13 0x0000000000429344 in request_handler_thread (arg=0x9333b0)
    at src/main/threads.c:690
        self = 0x9333b0
#14 0x00007ffff7342184 in start_thread (arg=0x7ffff0b0b700)
    at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7ffff0b0b700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737231501056,
                7779334135500148084, 1, 0, 140737231501760, 140737231501056,
---Type <return> to continue, or q <return> to quit---
                -7779301722961310348, -7779316046496494220},
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0},
            data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#15 0x00007ffff6e3637d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 3 (Thread 0x7ffff1c37700 (LWP 12637)):
#0  sem_wait () at ../nptl/sysdeps/unix/sysv/linux/x86_64/sem_wait.S:85
No locals.
#1  0x00000000004290af in request_handler_thread (arg=0x930b90)
    at src/main/threads.c:621
        self = 0x930b90
#2  0x00007ffff7342184 in start_thread (arg=0x7ffff1c37700)
    at pthread_create.c:312
        __res = <optimized out>
        pd = 0x7ffff1c37700
        now = <optimized out>
---Type <return> to continue, or q <return> to quit---
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140737249507072,
                7779334135500148084, 1, 9346240, 140737249507776,
                140737249507072, -7779303516110156428, -7779316046496494220},
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0},
            data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#3  0x00007ffff6e3637d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
No locals.

Thread 1 (Thread 0x7ffff7fec780 (LWP 12631)):
#0  0x00007ffff6e2dc53 in select () at ../sysdeps/unix/syscall-template.S:81
No locals.
#1  0x00007ffff79947bc in fr_event_loop (el=0x8ac6f0) at src/lib/event.c:600
        i = <optimized out>
        rcode = <optimized out>
        when = {tv_sec = 0, tv_usec = 211319}
        wake = 0x7fffffffe410
        maxfd = 17
---Type <return> to continue, or q <return> to quit---
        read_fds = {fds_bits = {260608, 0 <repeats 15 times>}}
        master_fds = {fds_bits = {260608, 0 <repeats 15 times>}}
#2  0x0000000000431651 in radius_event_process () at src/main/process.c:5622
No locals.
#3  0x000000000040bbd6 in main (argc=<optimized out>, argv=0x7fffffffe6b8)
    at src/main/radiusd.c:585
        rcode = 0
        status = <optimized out>
        argval = <optimized out>
        spawn_flag = true
        display_version = false
        flag = 0
        from_child = {-1, -1}
        p = <optimized out>
        state = 0x657840 <global_state>
        autofree = 0x658070
```

**Comment by arr2036** [comment]
Have you tried with v3.0.x head?

**Comment by arr2036** [comment]
and could you provide the full backtrace, instead of just starting GDB...

**Comment by pyrrhudite** [comment]
Sorry about that, I just deployed the freeradius stable v3.0.11, and the "info treads" has been updated.

**Comment by arr2036** [comment]
Can you please deploy v3.0.x HEAD. i.e.
```bash
git clone git@github.com:FreeRADIUS/freeradius-server.git
cd freeradius-server
git checkout v3.0.x
./configure ...etc
```

**Comment by pyrrhudite** [comment]
well, thanks for your advice. The problem seems fixed in v3.0.x HEAD.
